src/scenario/scenario_events_controller.c

```diff
@@ -0,0 +1,253 @@
+#include "scenario/condition_types/condition_handler.h"
+#include "scenario/scenario.h"
+#include "scenario/scenario_event.h"
+
+#define SCENARIO_EVENTS_SIZE_STEP 50
+
+static array(scenario_event_t) scenario_events;
+
+static int event_in_use(const scenario_event_t *event)
+{
+    return event->state != EVENT_STATE_UNDEFINED;
+}
+
+void scenario_events_init(void)
+{
+    for (int i = 0; i < scenario_events.size; i++) {
+        scenario_event_t *current = array_item(scenario_events, i);
+        scenario_event_init(current);
+    }
+}
+
+void new_scenario_event(scenario_event_t *obj, int position)
+{
+    obj->id = position;
+}
+
+void scenario_events_clear(void)
+{
+    scenario_events.size = 0;
+    if (!array_init(scenario_events, SCENARIO_EVENTS_SIZE_STEP, new_scenario_event, event_in_use)) {
+        log_error("Unable to allocate enough memory for the scenario events array. The game will now crash.", 0, 0);
+    }
+}
+
+scenario_event_t *scenario_event_get(int event_id)
+{
+    return array_item(scenario_events, event_id);
+}
+
+scenario_event_t *scenario_event_create(int repeat_min, int repeat_max, int max_repeats)
+{
+    if (repeat_min > 0 && repeat_max < repeat_min) {
+        log_error("Event maximum repeat is less than its minimum.", 0, 0);
+        return 0;
+    }
+
+    scenario_event_t *event = 0;
+    array_new_item(scenario_events, 0, event);
+    if (!event) {
+        return 0;
+    }
+    event->state = EVENT_STATE_ACTIVE;
+    event->repeat_months_min = repeat_min;
+    event->repeat_months_max = repeat_max;
+    event->max_number_of_repeats = max_repeats;
+
+    return event;
+}
+
+void scenario_event_delete(scenario_event_t *event)
+{
+    memset(event, 0, sizeof(scenario_event_t));
+    event->state = EVENT_STATE_UNDEFINED;
+    array_trim(scenario_events);
+}
+
+int scenario_events_get_count(void)
+{
+    return scenario_events.size;
+}
+
+static void scenario_events_info_save_state(buffer *buf)
+{
+    int32_t array_size = scenario_events.size;
+    int32_t struct_size = (6 * sizeof(int32_t)) + (3 * sizeof(int16_t));
+    buffer_init_dynamic_piece(buf,
+        SCENARIO_EVENTS_VERSION,
+        array_size,
+        struct_size);
+
+    for (int i = 0; i < scenario_events.size; i++) {
+        scenario_event_save_state(buf, array_item(scenario_events, i));
+    }
+}
+
+static void scenario_events_conditions_save_state(buffer *buf)
+{
+    int32_t array_size = 0;
+    for (int i = 0; i < scenario_events.size; i++) {
+        scenario_event_t *current = array_item(scenario_events, i);
+        array_size += current->conditions.size;
+    }
+    // If in future conditions can be linked to more things, then also take them into account here.
+
+    int32_t struct_size = (2 * sizeof(int16_t)) + (6 * sizeof(int32_t));
+    buffer_init_dynamic_piece(buf,
+        SCENARIO_EVENTS_VERSION,
+        array_size,
+        struct_size);
+
+    for (int i = 0; i < scenario_events.size; i++) {
```

Comment:
Just want to make sure I understand the array_foreach define correctly.
I could still use it for the outer for loop, correct?
Since it is just the equivalent of the code it defines array_foreach as?

Reply:
Yes, you can use it for the outer loop.